### DISCUSSION TITLE
Paper Reading Agent Integration (see Sakana AI)

### DISCUSSION THREAD

**Discussion by BradKML** [SmartManoj/Kevin#136] [Ideas]
Saw this repo recently, and that it is very inspiring that such agents can automatically dig papers and find new ideas to test, but they used Aider instead of OpenHands, and cannot write code out of scratch. Maybe some of the ideas can be borrowed to make adding features to other codebases easier? https://github.com/SakanaAI/AI-Scientist

**Comment by SmartManoj** [comment]
Added [download_arxiv_pdf](https://github.com/SmartManoj/Kevin/commit/3730ec7200ba2833a759c92e70e5828ddb1e851a) skill.

Aider [can write code](https://aider.chat/docs/scripting.html#python) out of scratch.  

The agent can use the [AI-Scientist](https://github.com/SakanaAI/AI-Scientist) repo directly for containerization.

**Reply by BradKML** [reply]
Could you also include other places that can get scientific paper? I think what makes AI-Scientist special is that it does not just include other sources of information, but also it is able to synthesize and create new ideas.

Some of the inspirations for the crawl-and-cite goal:
- GPT Researcher (includes autonomous web browsing) https://github.com/assafelovic/gpt-researcher
- https://github.com/TheBlewish/Automated-AI-Web-Researcher-Ollama
- https://github.com/stanford-oval/storm

**Reply by SmartManoj** [reply]
[Integrated](https://github.com/SmartManoj/Kevin/commit/07a6e6e09deec59527dab824a9ad45b77cd57ced) web browser using [undetected-chromedriver](https://github.com/ultrafunkamsterdam/undetected-chromedriver) and [semantic scholar](https://github.com/danielnsilva/semanticscholar)

**Reply by BradKML** [reply]
@SmartManoj regarding web crawling I have another place to park this https://github.com/SmartManoj/Kevin/discussions/138

As for paper hunting are Google Scholar, ResearchGate, CiteSeerx, tandfonline, and academia.edu any good as a private service?
What about Open Access aids like OpenAccessButton, PaperPanda, OA.mg, UnPaywall, DOAJ, SmartQuantAI, and core.ac.uk?

**Reply by SmartManoj** [reply]
Should AI Agents solve captchas? 
https://www.researchgate.net/post/API_on_ResearchGatenet

---
~Integrating~ https://github.com/EsdrasCaleb/resp

**Reply by SmartManoj** [reply]
[Integrated](https://github.com/SmartManoj/Kevin/commit/709d02701764e1c98854b9e27fa02611a4bf4db3) Google Scholar using [scholarly](https://github.com/scholarly-python-package/scholarly). Most of the papers are available and can be manually added. So, do you think there is no need to add other providers?

**Reply by BradKML** [reply]
Currently, the ideal would be to prioritize open access sites since a lot of them might not be covered by fetchers. Some of the OA sites have wider coverage than Arxiv and Google Scholar so those need attention as well (use paper count as reference).
As for the private services like ResearchGate, it would be good to note which ones has standard APIs to apply for, and which ones do not first. Optimistic about academia.edu tho.

Side note: we should **be very cautious** about the "alternatives" like Anna's Archive and other SLUM sources
- "Master source" https://annas-archive.org
- Source directories https://open-slum.org 
- Books https://libgen.st https://z-library.sk/
- Academic Papers https://sci-hub.ru/

**Reply by SmartManoj** [reply]
Currently, Is there any paper that is not found in Google Scholar but in academia.edu?

[A Perplexity Comparison](https://www.perplexity.ai/search/google-scholar-vs-academia-edu-BOIivU8aSjylg.lulyzafg)

**Reply by BradKML** [reply]
I think [CiteSeerx would have exclusives while Academia.edu has some "indie titles"](https://www.perplexity.ai/search/which-of-the-following-has-the-FCJi31_aTRK1M3XkZJjIOA)
And for Open Access, OA.mg and core.ac.uk for quantity, and then Unpaywall and DOAJ for niche papers.
There are also OpenAIRE, UKRI, HAL, and BASE, among other institution-centric repos, listed under the OpenDOAR superlist. But those are secondary

**Reply by SmartManoj** [reply]
Added [download_academia_pdf](https://github.com/SmartManoj/Kevin/commit/65d0ab7bc44d823ea2501beb9b48a3371411741f) skill

**Reply by BradKML** [reply]
Definitely worth checking out these ones at least according to the table (all of them have more than SemanticScholar) https://oa.mg/data
- OA.mg https://chrome.google.com/webstore/detail/oamg/ahnnebfmpaambhblikbbbhdfaknfjfcd
- IA Scholar https://github.com/internetarchive/fatcat https://github.com/internetarchive/fatcat-scholar
- UnPaywall https://github.com/michaelplews/python-unpaywall
- CORE.ac.uk (a bit less and closer to SemanticScholar) https://github.com/ronentk/sci-paper-miner

**Reply by SmartManoj** [reply]
Added support for OA.mg and integrated [PyPaperBot](https://github.com/ferru97/PyPaperBot) which is again being developed. 
Support for other websites should be added to that library for best of both worlds.

**Reply by BradKML** [reply]
Also looking at how PyPaperBot is different from the other major repos https://github.com/jonatasgrosman/findpapers

**Reply by SmartManoj** [reply]
Agent [downloaded](https://www.all-hands.dev/share?share_id=169bb91a3790677897e8c817d799f3e42fc0d8e93506b8c25013ce07cd5ea9e8) a paper using Selenium.

**Reply by BradKML** [reply]
Ideally, in order of priority:
1. Download/Scrape/Crawl/Fetch scientific/academic papers with established packages (e.g. PyPaperBot) for main sites (select top sites with most papers and those with the most rare papers)
2. Modify established packages to accommodate other paper sites based on API (or take inspiration from less popular repositories), and note if the paper APIs are rate-limited or not, make PRs if necessary
3. Use Playwright/Selenium/Appium for Python, or Puppeteer/Cypress/Jest/Mocha/Karma/Nightwatch for JS/TS, to download papers, and make sure that the methodology is stable and usable

**Reply by erkinalp** [reply]
try https://github.com/SakanaAI/AI-Scientist-v2

**Comment by BradKML** [comment]
As a sidenote, I need to remind myself that this exists. Paper hints as lacking in (in descending order of severity) specificity, human-level helpfulness, and alignment. It had issues in communicating novelty, citations, and theoretical soundness (and for ML specifically, ablation studies), while overemphasizing broadening datasets and elaborating on implications. https://github.com/Weixin-Liang/LLM-scientific-feedback
LLM needs to be finetuned/selected to giving better feedback (and not overfit), yet it also requires human data to do so.

**Reply by BradKML** [reply]
Some extra hints about CoT-free LLMs:
- LLMs can only catch errors like half of the time using defaults in GPT-4, can't even tell which papers are better than the others consistently, but can absolutely work checklists https://doi.org/10.48550/arXiv.2306.00622
- LLMs (short context window GPT) can read reviews accurately BUT they cannot read papers, the best bet is using chunk selection https://aclanthology.org/2024.lrec-main.816
- LLMs can't help with identifying reviewer's unfounded statements, inaccurateness, superficiality, lack of clarity, self-contradiction, and low experiment quality (and to a lesser extent, copy-paste or short summaries for non-Gemini closed weight models)... while also are more likely to have review issues with scope creep, lack of clarity, superficiality, self-contradiction, unfounded statements, and duplication of content. https://arxiv.org/html/2406.16253v3

Could CoTs help with (a) lack of clarity and superficiality through technique change, (b) contradiction and unfounded statements through self-reflection, (c) ranking papers through section-based reading?

**Reply by SmartManoj** [reply]
How do you rate [this Paper feedback](https://telegra.ph/Gemini-Response-12-21)?

---

> LLMs can only catch errors like half of the time using defaults in GPT-4, 

Auto-correcting the error, and iterating until no error would solve this.

**Reply by BradKML** [reply]
Would like to dig into those papers again, and then observe some of the issues. Blind auto-correction might be useless if the issues are persistent and more systematic (as in failing as both reviewer and meta-reviewer).
But in those experiments things like "Chain-of-Thought" (e.g. self-consistency, beam search, socratic reasoning, graph-of-thought) are not yet tried before. Self-contradiction and unfounded statements can be fixed in that regard, so some of the mechanisms need to be explored.
I am not sure about superficiality and vagueness tho, those seem to require more subtle ways of auto-correction, for some reason Chain-of-Density could be a buzzword cure (recursive summarization for article sections) but also I have seen people discuss that this is purely an issue of prompt engineering.
LLMs can fix scope creep and duplication of content upon naive self-reflection (review failure) so it is trivial, while also having no issues dodging inaccurateness and low experiment quality without knowing why it is the case (meta-review failure) which makes rewriting viable in some cases.

**Reply by SmartManoj** [reply]
[Fragment](https://www.prompthub.us/blog/better-summarization-with-chain-of-density-prompting#:~:text=Density%20step%20was-,step%203,-%2C%20indicating%20a%20balance)
![image](https://github.com/user-attachments/assets/6e0c2e27-a83b-49e8-94b7-a1b78658a25b)


Here, should one prefer CoD at Step 2 and isn't the median preferred step 4?

**Reply by BradKML** [reply]
1 for coherence, 2 for quality/preference, 3 for attributable, 4 for informative/overall. Maybe doing all 4 extracts the most information based on different attribute?

**Reply by SmartManoj** [reply]
This is Human evaluation table

**Reply by BradKML** [reply]
I know, and I am thinking how majority support 2-step (corresponds to AI eval as "quality"), even when the median as 3-step (corresponding to "attributable"), and the maximum is 4-step (one of the researcher preferred "information"). As you can see each of the researcher has their own biases in what they want (quality vs attributable vs information). So for conservative estimate, anything with more than 4-step is likely going to be too cumbersome to read, with 2-step and 3-step as an optional.

**Reply by SmartManoj** [reply]
```python
# Percentages for Step 1 by individual annotators
individual_annotators = [3.0, 2.0, 13.0, 17.4]

# Aggregate calculation as the average of all annotators' percentages
aggregate = sum(individual_annotators) / len(individual_annotators)

print(f"Aggregate for Step 1: {aggregate:.2f}%")
```
1) Aggregate for Step 1: 8.85%?

---

2) Why are they taking the median preferred?

---

3) For GPT-4 Evaluation, first, what do you think about [9.9 vs 9.11 comparison](https://chatgpt.com/share/67690499-4cb4-8010-9cb4-517bbe3c33a4)?

---

Step 2 is best based on both evaluation here.

**Reply by BradKML** [reply]
1. Think there might be some rounding errors + irregularities
2. Median is better than the mean due to how sometimes distribution can be skewed towards the majority (which is 2-step). But the first and second rater in the table is already enough to draw some suspicion that 2 is not enough in some cases.
3. GPT-4 definitely fails in contextuality, when it is number it fails, but when it is software versioning it succeeds. Same is said regarding "number of Rs in strawberry" if it is spelling correction it did with in 2 Rs (counter is like "library", "accessary", "arbitrary", "dietary", "military", "ordinary", "sanitary"). Context matters in these cases.

**Reply by SmartManoj** [reply]
They should choose the mode instead of mean or median. The human annotators thought Step 2 was best for 30.8% of samples.

**Comment by BradKML** [comment]
As for paper crawling and recommendation services, Thinking which ones would be better when integrated based on price, accessibility, and function.

**Reply by BradKML** [reply]
@SmartManoj apart from just plain paper search and fetching, finding related papers (rather than keyword based search) are also an issue, maybe this should be a separate issue altogether? Some are more open than others tho.

**Reply by SmartManoj** [reply]
[Semantic Scholar](https://www.semanticscholar.org/) is a free, **AI-powered** research tool for scientific literature.

**Reply by BradKML** [reply]
SemanticScholar mainly focuses on the search engine side of the equation and not the related paper hunting side (with topical networks/graphs), so not sure how they can be related. It's a little bit reaching tho

**Reply by SmartManoj** [reply]
Could you provide links to a Paper and one of its related papers?

**Reply by BradKML** [reply]
Here is an example of how graph-centric tools look like, where topics and similar papers can be extracted https://semopenalex.org/resource/semopenalex:UniversalSearch
Will also make a list on the others later just in case

**Reply by BradKML** [reply]
@SmartManoj extra note: someone made a blog post on "literature mapping" and it could be of interest https://musingsaboutlibrarianship.blogspot.com/p/list-of-innovative-literature-mapping.html
There are three forms: (a) Citation network (bibliographic coupling and co-citation), (b) Co-authorship graph, and (c) keyword similarities. Thinking which service would be the most useful first

**Reply by SmartManoj** [reply]
https://www.connectedpapers.com/

**Comment by BradKML** [comment]
I will need to add paper evaluation into the reading process as well. Originally SakanaAI related but also applies to filtering papers and possibly even books and webpages if the user has preferences, or there is a more universal benchmark on "quality" https://github.com/Weixin-Liang/LLM-scientific-feedback/issues/32

**Comment by BradKML** [comment]
V2 came out https://github.com/SakanaAI/AI-Scientist-v2
